File: LeetCode/459. Repeated Substring Pattern/Solution.java
public class Solution {
    public boolean repeatedSubstringPattern(String str) {
        char[] cs = str.toCharArray();
        int n = cs.length;
        for (int i = 1; i <= n / 2; i++) {
            if (n % i != 0) {
                continue;
            }
            int d = n / i;
            if (check(cs, i, d)) {
                return true;
            }
        }
        return false;
    }

    private boolean check(char[] cs, int i, int d) {
        for (int j = 1; j < d; j++) {
            for (int k = 0; k < i; k++) {
                if (cs[i * j + k] != cs[k]) {
                    return false;
                }
            }
        }
        return true;
    }

    public boolean repeatedSubstringPattern1(String str) {
        return (str.substring(1) + str.substring(0, str.length() - 1)).indexOf(str) != -1;
    }
}

File: LeetCode/459. Repeated Substring Pattern/Solution.py
class Solution(object):
    def repeatedSubstringPattern(self, str):
        """
        :type str: str
        :rtype: bool
        """
        n = len(str)
        for i in range(1, n // 2 + 1):
            if n % i != 0:
                continue
            d = n // i
            for j in range(1, d):
                if str[0:i] != str[j * i:(j + 1) * i]:
                    break
            else:
                return True
        return False

    def repeatedSubstringPattern1(self, str):
        return str in (str+str)[1:-1]

